### PR TITLE
Move binary context to request/response header

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR moves the binary contexts from the frame trailer to the frame header.

It also:
- refactors outgoing frame marshaling to marshal the frame header only upon "send"
- simplifies OutgoingFrame and IncomingFrame: they now only hold the payload buffer (no header or trailer)
- updates OutputStream.Finish() to really finish the underlying buffer (could not do this previously as we could write the binary context after the encapsulation)
- changes the coloc invocation / dispatch logic to no longer fully marshal/unmarshal the frames. The frame header (with the exception of binary context) is no longer marshaled/unmarshaled.
- renames the IncomingFrame and OutgoingFrame Size to PayloadSize (this has little impact on responses since until this PR they didn't have a header; however, this reduces the traced/observed size for requests.

### CODE REVIEW

File: csharp/src/Ice/OutgoingFrame.cs [hunk 1]
```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
...
+                    OutputStream.Position startValue = ostr.StartFixedLengthSize(2);
+                    action(ostr);
+                    ostr.EndFixedLengthSize(startValue, 2);
+                    size++;
                 }
-                else
+            }
+            foreach ((int key, ReadOnlyMemory<byte> value) in InitialBinaryContext)
+            {
+                if (_binaryContextOverride == null || !_binaryContextOverride.ContainsKey(key))
```
Comment (pepone) [inline csharp/src/Ice/OutgoingFrame.cs:239]:
```suggestion
                if (!(_binaryContextOverride?.ContainsKey(key) ?? false))
```

File: csharp/src/Ice/SocketStream.cs [hunk 2]
```diff
@@ -517,14 +519,25 @@ internal void TryDispose()
...
-            CancellationToken cancel)
+        private protected virtual async ValueTask SendFrameAsync(OutgoingFrame frame, CancellationToken cancel)
         {
             // The default implementation only supports the Ice2 protocol
             Debug.Assert(_socket.Endpoint.Protocol == Protocol.Ice2);
 
-            if (frame.Size > _socket.PeerIncomingFrameMaxSize)
+            var buffer = new List<ArraySegment<byte>>(frame.Payload.Count + 1);
+            var ostr = new OutputStream(Encoding.V20, buffer);
+            ostr.WriteByteSpan(Header.Span); // TODO: rename Header to TransportHeader?
```
Comment (bentoi) [inline csharp/src/Ice/SocketStream.cs:529]: Renaming sounds good!

File: csharp/src/Ice/OutgoingFrame.cs [hunk 3]
```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
...
             }
         }
 
+        /// <summary>Gets or builds a combined binary context using InitialBinaryContext and _binaryContextOverride.
+        /// This method is used for colocated calls.</summary>
+        internal IReadOnlyDictionary<int, ReadOnlyMemory<byte>> GetBinaryContext()
+        {
+            if (_binaryContextOverride == null)
+            {
+                return InitialBinaryContext;
```
Comment (pepone) [inline csharp/src/Ice/OutgoingFrame.cs:181]: Is it safe to return the InitialBinaryContext here? the incoming and outgoing frames will share the ROM for these entries, you can not release it after sent, probably not very important as this code path is only used for collocated

**Review by pepone — APPROVED** [review]
Looks good, just a few comments

File: csharp/src/Ice/OutgoingFrame.cs [hunk 4]
```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
...
+    }
 
-        private protected bool ContainsKey(int key) => _binaryContextKeys?.Contains(key) ?? false;
-
-        private bool AddKey(int key)
-        {
-            _binaryContextKeys ??= new HashSet<int>();
-            return _binaryContextKeys.Add(key);
-        }
+    public static class OutgoingFrameHelper
```
Comment (bentoi) [inline csharp/src/Ice/OutgoingFrame.cs:251]: What is the purpose of this public method? It's not clear how users will be able to use it and for what purpose.

**Review by bentoi — APPROVED** [review]
Looks fine, added few comments.

File: csharp/src/Ice/BinaryContext.cs [hunk 5]
```diff
@@ -1,11 +1,43 @@
...
-    public enum BinaryContext : int
+    public enum BinaryContextKey : int
     {
         /// <summary>Key used to encode the retry policy entry.</summary>
         RetryPolicy = -1
     }
+
+    public static class BinaryContextHelper
+    {
+        /// <summary>Reads a binary context from the steam.</summary>
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/BinaryContext.cs:18]: Fixed.

File: csharp/src/Ice/Ice1Definitions.cs [hunk 6]
```diff
@@ -75,12 +52,15 @@ internal enum FrameType : byte
...
                 Magic[0], Magic[1], Magic[2], Magic[3],
                 ProtocolBytes[0], ProtocolBytes[1], ProtocolBytes[2], ProtocolBytes[3],
-                (byte)FrameType.ValidateConnection,
+                (byte)Ice1FrameType.ValidateConnection,
                 0, // Compression status.
                 HeaderSize, 0, 0, 0 // Frame size.
             }
         };
 
+        private static readonly byte[] _voidReturn11 = new byte[] { 0, 6, 0, 0, 0, 1, 1 };
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Ice1Definitions.cs:61]: Maybe _voidReturnPayload11 (etc)?
We use "void return" and "empty args" consistently for other methods too.

File: csharp/src/Ice/OutgoingFrame.cs [hunk 7]
```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
...
+    {
+        /// <summary>Writes the header of a frame. This header does not include the frame's prologue.</summary>
+        /// <param name="ostr">The output stream.</param>
+        /// <param name="frame">The frame.</param>
+        public static void WriteHeader(this OutputStream ostr, OutgoingFrame frame) =>
+            frame.WriteHeader(ostr);
     }
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingFrame.cs:258]: This extension provides the usual `ostr.Write(...)` or `ostr.WriteXxx(...)` public API.

File: csharp/src/Ice/Ice2Definitions.cs [hunk 8]
```diff
@@ -12,29 +12,41 @@ internal static class Ice2Definitions
...
             IReadOnlyList<string> location,
             string operation,
             bool idempotent,
-            DateTime deadline)
+            DateTime deadline,
+            IReadOnlyDictionary<string, string> context)
         {
             Debug.Assert(ostr.Encoding == Encoding);
-            BitSequence bitSequence = ostr.WriteBitSequence(4); // bit set to true (set) by default
+            BitSequence bitSequence = ostr.WriteBitSequence(5); // bit set to true (set) by default
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Ice2Definitions.cs:49]: Fixed.

File: csharp/src/Ice/IncomingFrame.cs [hunk 9]
```diff
@@ -12,76 +11,43 @@ namespace ZeroC.Ice
...
-                        _binaryContext = ImmutableDictionary<int, ReadOnlyMemory<byte>>.Empty;
-                    }
-                }
-                return _binaryContext;
-            }
-        }
-        /// <summary>The encoding of the frame payload.</summary>
-        public abstract Encoding Encoding { get; }
-        /// <summary>True if the encapsulation has a compressed payload, false otherwise.</summary>
+        /// <summary>Returns true this payload holds a compressed encapsulation payload; otherwise, returns false.
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/IncomingFrame.cs:17]: Fixed.

File: csharp/src/Ice/OutgoingRequestFrame.cs [hunk 10]
```diff
@@ -322,77 +300,69 @@ public void Dispose()
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingRequestFrame.cs:300]: Fixed, thanks.

File: csharp/src/Ice/OutgoingFrame.cs [hunk 11]
```diff
@@ -23,125 +22,96 @@ public enum CompressionResult
     /// <summary>Base class for outgoing frames.</summary>
     public abstract class OutgoingFrame
     {
-        /// <summary>The encoding of the frame payload.</summary>
-        public abstract Encoding Encoding { get; }
+        /// <summary>Returns a dictionary used to override the binary context of this frame. The full binary context
+        /// is a combination of <see cref="InitialBinaryContext"/> plus these overrides.</summary>
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingFrame.cs:26]: Fixed.

File: csharp/src/Ice/OutgoingFrame.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingFrame.cs:239]: I find the suggested version harder to understand!

File: csharp/src/Ice/BinaryContext.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/BinaryContext.cs:18]: Fixed.

File: csharp/src/Ice/IncomingFrame.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/IncomingFrame.cs:17]: Fixed.

File: csharp/src/Ice/SocketStream.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/SocketStream.cs:529]: Renamed and removed TODO.

File: csharp/src/Ice/OutgoingFrame.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingFrame.cs:251]: This public extension method allows users to marshal the header of a frame.
I'll remove it for now, we can add it back if we find a use-case.

File: csharp/src/Ice/OutgoingFrame.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingFrame.cs:258]: As per Benoit's comment, I'll remove this extension.

File: csharp/src/Ice/OutgoingResponseFrame.cs [hunk 12]
```diff
@@ -53,7 +54,7 @@ public static OutgoingResponseFrame WithVoidReturnValue(Current current)
         {
             (OutgoingResponseFrame response, OutputStream ostr) = PrepareReturnValue(current, compress, format);
             writer(ostr, returnValue);
-            response.PayloadEnd = ostr.Finish();
+            _ = ostr.Finish();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingResponseFrame.cs:57]: Fixed as suggested. We were using the return only in one place, for an assert.

File: csharp/src/Ice/OutgoingFrame.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingFrame.cs:181]: Currently it's safe, but we'll need to be careful when we implement array pooling, and the solution then could be to make a copy of this memory.

File: csharp/src/Ice/ColocatedStream.cs [hunk 13]
```diff
@@ -82,10 +81,104 @@ internal void ReceivedFrame(object frame, bool fin)
             {
                 // Run the continuation asynchronously if it's a response to ensure we don't end up calling user
                 // code which could end up blocking the AcceptStreamAsync task.
-                SignalCompletion((frame, fin), runContinuationAsynchronously: frame is OutgoingResponseFrame);
+                SignalCompletion((frame, fin), runContinuationAsynchronously: frame is IncomingResponseFrame);
             }
         }
 
+        internal override async ValueTask<IncomingRequestFrame> ReceiveRequestFrameAsync(CancellationToken cancel)
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ColocatedStream.cs:88]: I added OutgoingFrame.ToIncoming() and replaced the SendXxxFrameAsync overrides by a single SendFrameAsync override, as suggested.

However, this does not remove the ReceiveXxxFrameAsync overrides. I let you simplify them in a follow-up PR!

File: csharp/src/Ice/ColocatedStream.cs [hunk 14]
```diff
@@ -82,10 +81,104 @@ internal void ReceivedFrame(object frame, bool fin)
...
+        {
+            (object frameObject, bool fin) = await WaitSignalAsync(cancel).ConfigureAwait(false);
+            var frame = (IncomingRequestFrame)frameObject;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ColocatedStream.cs:91]: Done.

File: csharp/src/Ice/ColocatedStream.cs [hunk 15]
```diff
@@ -82,10 +81,104 @@ internal void ReceivedFrame(object frame, bool fin)
...
+            catch (OperationCanceledException)
+            {
+                if (_socket.Endpoint.Protocol != Protocol.Ice1)
+                {
+                    await ResetAsync((long)StreamResetErrorCode.RequestCanceled).ConfigureAwait(false);
+                }
+                throw;
+            }
+
+            var frame = (IncomingResponseFrame)frameObject;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ColocatedStream.cs:123]: Done.

File: csharp/src/Ice/OutgoingRequestFrame.cs [hunk 16]
```diff
@@ -64,16 +45,25 @@ public sealed class OutgoingRequestFrame : OutgoingFrame, IDisposable
         /// <summary>The operation called on the Ice object.</summary>
         public string Operation { get; }
 
-        private SortedDictionary<string, string>? _contextOverride;
-        private readonly ArraySegment<byte> _defaultBinaryContext;
+        /// <inheritdoc/>
+        public override Encoding PayloadEncoding { get; }
+
+        /// <summary>WritableContext is writable version of Context. Its entries are always the same as Context's
```
Comment (externl) [inline csharp/src/Ice/OutgoingRequestFrame.cs:51]:
```suggestion
        /// <summary>WritableContext is a writable version of Context. Its entries are always the same as Context's
```

File: csharp/src/Ice/IncomingFrame.cs [hunk 17]
```diff
@@ -90,70 +56,59 @@ public void DecompressPayload()
...
+                        @$"decompressed size of {decompressedSize
+                        } bytes is greater than the configured IncomingFrameMaxSize value ({_maxSize} bytes)");
                 }
 
-                // We are going to replace the Data segment with a new Data segment/array that contains a decompressed
-                // encapsulation.
-                byte[] decompressedData = new byte[Data.Count - size + decompressedSize];
+                // We are going to replace the Payload segment with a new Payload segment/array that contains a
+                // decompressed encapsulation.
+                byte[] decompressedPayload = new byte[Payload.Count - size + decompressedSize];
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/IncomingFrame.cs]: We should not make this assumption. decompressedSizeLength could for example be `2` while sizeLength is `4`.

I updated the code which should be more correct now. Please re-review.

In a follow-up PR, we should improve the handling of unknown compression algorithms. We currently assume != 0 means gzip, which is not future proof. Maybe something like:
```
// Slice
unchecked enum PayloadCompressionStatus
{
    NotCompressed = 0,
    GZip = 1
}
```

```
// C#
public abstract class IncomingFrame
{
    public PayloadCompressionStatus PayloadCompressionStatus { get; } // replaces HasCompressedPayload
}
```

and `DecompressPayload` would throw NotSupportedException when attempting the decompression a payload with an unknown compression status.

**Review by externl — APPROVED** [review]
Looks good to me

File: csharp/src/Ice/OutgoingRequestFrame.cs (hunk 16 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutgoingRequestFrame.cs:51]: Fixed.

**Review by pepone — APPROVED** [review]
Looks good!

**Pull request merged**